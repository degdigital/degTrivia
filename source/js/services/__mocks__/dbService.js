const dbService = jest.genMockFromModule('./dbService');

function submitAnswer() {
	return Promise.resolve(true);
}

function getLeaderboardData() {
	return Promise.resolve({
		game: [
			{name: 'Anna', score: '1500000'},
			{name: 'Aaron', score: '140'},
			{name: 'Ryan', score: '42'}
		],
		event: [
			{name: 'Aaron', score: '1500001'},
			{name: 'Aaron', score: '200'},
			{name: 'Ryan', score: '43'}
		]
	})
}

dbService.submitAnswer = submitAnswer;
dbService.getLeaderboardData = getLeaderboardData;

export default dbService;