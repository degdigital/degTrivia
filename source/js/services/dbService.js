import firebase from '@firebase/app';
import '@firebase/database';

const dbService = function() {

	let db = null;
	const nodeNames = {
		events: 'events',
		series: 'series',
		games: 'games',
		pendingPlayers: 'pendingPlayers'
	};
	const defaultNodesToGet = [
		'events',
		'series',
		'games'
	];

	function init() {
		db = firebase.database();
	}

	function getDb() {
		return db;
	}

	function getInitialData(nodes = defaultNodesToGet) {
		return new Promise(async(resolve, reject) => {
			let output = {};
			const data = await Promise.all(nodes.map(node => db.ref(node).once('value').then(snapshot => snapshot.val())));
			nodes.map((node, index) => output[node] = data[index]);
			resolve(output);
		});
	}

	function getCurrentEventId() {
		return db.ref('currentEvent').once('value').then(snapshot => snapshot.val());
	}

	function getEvent(eventAlias) {
		return db.ref('events').orderByChild('alias').equalTo(eventAlias).once('value').then(snapshot => snapshot.val());
	}

	function createPendingPlayer(playerVals, eventKey, userId) {
		const pendingPlayerKey = db.ref(nodeNames.pendingPlayers).push().key;
		
		return db.ref(`/${nodeNames.pendingPlayers}/${pendingPlayerKey}`).update(formattedPlayerVals)
			.then(() => pendingPlayerKey);
	}

	async function getNextGameTime() {
		const currentEventId = await getCurrentEventId();
		const games = await db.ref('games').orderByChild('event').equalTo(currentEventId).once('value').then(snapshot => snapshot.val());
		if (!games) {
			return Promise.resolve(null);
		}
		const nextGameTime = new Date(Object.keys(games).map(gameId => games[gameId].startTime).sort()[0]);
		return Promise.resolve(nextGameTime);
	}

	function submitAnswer(questionId, choiceId, playerId) {
		if (questionId && choiceId && playerId){
			return db.ref(`answers/${questionId}/responses/${choiceId}`).update({
				[playerId]: true
			});
		}
	}

	function getActiveGameData(gameId) {
		return new Promise((resolve, reject) => {
			db.ref(`games/${gameId}`).once('value', snapshot => {
				if (snapshot.exists()) {
					resolve(snapshot.val());	
				} else {
					reject();
				}
			});
		});
	}

	function getLeaderboardData() {
		return db.ref('leaderboardCurrent').once('value').then(snapshot => {
			if (snapshot.val()) {
				return snapshot.val();
			} else {
				return {};
			}
		})
	}

	return {
		init,
		getDb,
		getInitialData,
		getCurrentEventId,
		getEvent,
		getNextGameTime,
		submitAnswer,
		getLeaderboardData,
		getActiveGameData
	};

};

const instance = dbService();

export default instance;