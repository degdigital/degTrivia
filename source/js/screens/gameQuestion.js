import {replaceContent} from '../utils/domUtils.js';
import router from '../utils/router.js';
import dbService from '../services/dbService.js';
import playerService from '../services/playerService.js';
import countdown from '../components/countdown.js';

const cssClasses = {
	countdownContainer: 'countdown-container',
	choices: 'choices',
	choiceButton: 'choice-button',
	choiceButtonSelected: 'choice-button--selected'
}

function bindEventListeners(element, data) {
	const onClickBound = e => onClick(e, element, data);

	element.addEventListener('click', onClickBound);

	return onClickBound;
}

function unbindEventListeners(element, onClickBound) {
	if(onClickBound) {
		element.removeEventListener('click', onClickBound);
	}
}

function onClick(e, element, {questionData}) {
	if(e.target.matches(`.${cssClasses.choiceButton}`)) {
		const selectedChoiceId = e.target.dataset.id;
		const playerId = playerService.getAuth().currentUser.uid;
		dbService.submitAnswer(questionData.id, selectedChoiceId, playerId);
		updateChoices(element, questionData.choices, selectedChoiceId);
	}
}

function renderChoices(choices, selectedChoiceId) {
	return Object.keys(choices).reduce((html, key) => {
		const isSelected = key === selectedChoiceId;
		const buttonClasses = [cssClasses.choiceButton];
		const disabledAttr = selectedChoiceId ? 'disabled' : '';
		
		if(isSelected) {
			buttonClasses.push(cssClasses.choiceButtonSelected);
		} 

		const buttonHtml = `<button class="${buttonClasses.join(' ')}" data-id="${key}" ${disabledAttr}>${choices[key].text}</button>\n`;

		return html + buttonHtml;
	}, '');
}

function updateChoices(element, choices, selectedChoiceId) {
	const choicesEl = element.querySelector(`.${cssClasses.choices}`);
	replaceContent(choicesEl, renderChoices(choices, selectedChoiceId));
}

function renderScreen(element, data) {
	const {gameId, questionData} = data;

	const html = `
		<div>
			<h1>Question #${questionData.order + 1}</h1>
			<div class="countdown-container"></div>
			<p>${questionData.question}</p>
			<div class="${cssClasses.choices}">
				${renderChoices(questionData.choices)}
			</div>
		</div>`;

	replaceContent(element, html);
}

function onCountdownComplete(data) {
	data.resultsPending = true;
	router.route('gameQuestionResults', data);
}

function teardown(element, onClickBound, countdownInst) {
	unbindEventListeners(element, onClickBound);
	countdownInst.stop();
}

export default function({element}) {
	let onClickBound;
	let countdownInst;

	function render(data) {
		onClickBound = bindEventListeners(element, data);
	
		const duration = data.questionData.expirationTime - Date.now();

		renderScreen(element, data);
		
		countdownInst = countdown({
			containerElement: element.querySelector(`.${cssClasses.countdownContainer}`),
			format: 'mm:ss',
			onComplete: () => onCountdownComplete(data)
		});
		countdownInst.start(duration, 'milliseconds');
	}

	return {
		render,
		teardown: () => teardown(element, onClickBound, countdownInst)
	};
}