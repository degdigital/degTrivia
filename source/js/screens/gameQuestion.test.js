import gameQuestion from './gameQuestion';
import dbService from '../services/dbService';

jest.mock('../services/dbService');

let element;

const data = {
	gameId: 1,
	questionData: {
		id: 10,
		order: 0,
		question: 'Who is the most dashing Kansas City Royal of all time?',
		choices: {
			100: 'Bob Hamelin',
			101: 'Danny Tartabul',
			102: 'Steve Balboni' 
		}
	}
};

beforeEach(() => {
	element = document.createElement('div');
	document.body.appendChild(element);
});

describe('calling render()', () => {
	test('renders out a question and choices', async () => {
		await gameQuestion(element).render(data);
		expect(element).toMatchSnapshot();
	});
});

describe('selecting a choice', () => {
	test('calls the dbService submitAnswer() method', async () => {
		const submitAnswerSpy = jest.spyOn(dbService, 'submitAnswer');

		await gameQuestion(element).render(data);

		const firstChoiceButtonEl = element.querySelector('.choice-button');
		firstChoiceButtonEl.click();

		const expectedChoiceId = Object.keys(data.questionData.choices)[0];

		expect(submitAnswerSpy).toHaveBeenCalledTimes(1);
		expect(submitAnswerSpy.mock.calls[0]).toEqual([data.gameId, data.questionData.id, expectedChoiceId]);
	});

	test('disables the choice buttons', async () => {
		await gameQuestion(element).render(data);

		const firstChoiceButtonEl = element.querySelector('.choice-button');
		firstChoiceButtonEl.click();

		const choiceButtonsEls = [...element.querySelectorAll('.choice-button')];
		expect(choiceButtonsEls[0].disabled).toBe(true);
		expect(choiceButtonsEls[1].disabled).toBe(true);
		expect(choiceButtonsEls[2].disabled).toBe(true);
	});
});