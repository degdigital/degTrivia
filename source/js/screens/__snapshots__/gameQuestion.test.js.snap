// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render() renders out a question and choices 1`] = `
<div>
  
		
  <div>
    
			
    <h1>
      Question #1
    </h1>
    
			
    <div
      class="countdown-container"
    />
    
			
    <p>
      Who is the most dashing Kansas City Royal of all time?
    </p>
    
			
    <div
      class="choices"
    >
      
				
      <button
        class="choice-button"
        data-id="100"
      >
        Bob Hamelin
      </button>
      

      <button
        class="choice-button"
        data-id="101"
      >
        Danny Tartabul
      </button>
      

      <button
        class="choice-button"
        data-id="102"
      >
        Steve Balboni
      </button>
      

			
    </div>
    
		
  </div>
</div>
`;
